### PR TITLE
Issue 26 backend

### PR DESCRIPTION
**Pull request by XuechunHou**
1. this request includes modified implementation on **/friendrequest** handler, it now can handles incoming/outgoing foreign friendrequest 
2. it also contains a fix of the @validate_remote_server_authentication decorator.

### CODE REVIEW

**Comment by XuechunHou** [comment]
it does not contain a fix of the @validate_remote_server_authentication decorator anymore. there was no bug in the decorator.

**Comment by XuechunHou** [comment]
It also altered Node table to set foreign_server_username UNIQUE

File: web_server/social_distribution/utils/basic_auth.py [hunk 1]
```diff
@@ -54,11 +54,12 @@ def handle_incoming_request_or_request(view, request, test_func, realm="", *args
...
                 # Check if the credentials are valid for the host requesting them
                 if Node.objects.all().filter(foreign_server_username=uname).exists():
-                    entry = Node.objects.get(foreign_server_username=uname)
+                    entry = Node.objects.all().get(foreign_server_username=uname)
+
                     if check_password(passwd, entry.foreign_server_password):
                         return view(request, *args, **kwargs)
                     else:
-                        deny_response += " The provided password for your server authentication was invalid"
+                        deny_response += " The provided password for yours server authentication was invalid"
```
Comment (AustinGrey) [inline web_server/social_distribution/utils/basic_auth.py]: Why this change? This takes it from correct grammar to incorrect grammar. Please undo this change.

**Review by AustinGrey — CHANGES_REQUESTED** [review]
There is a grammatical issue in one of the changes.

Additionally I am unsure how to test this? Is it the same process as for our local server, except I just need to spin up a remote server as well?

File: web_server/social_distribution/utils/basic_auth.py (hunk 1 above)
Comment (XuechunHou) [reply web_server/social_distribution/utils/basic_auth.py]: i don't remember I changed that.  lol. anyway i will remove "s"
